[Ready] Belt and Holster Overhaul

CakeQ opened the pull request:
🆑 Cakey
rscadd: Belts have been overhauled to have holster slots on certain belts. These holsters function the exact same way as uniform holsters do.
imgadd: Certain belts now show what their contents are on their icons.
rscadd: Added general and general holster belts, used by command, bridge officers, and supply. Holds general office supplies, tablets, and stamps as well as other various supply equipment.
maptweak: Uniform holsters have been removed in favor of belt holsters, however they can still be found within the loadout menu.
tweak: Uniform holsters have been modified to use the same slot as webbings to counter being able to carry two holstered weapons.
/🆑

Holsters have been overhauled to become an extension, which can be attached to any obj/weapon/storage. Also added some QoL to make holsters more interactable based on intent. You can now view the holster's storage UI on help intent.

![image](https://user-images.githubusercontent.com/17550641/42405989-df8881dc-8196-11e8-8ac5-7e3171386188.png)

![image](https://user-images.githubusercontent.com/17550641/42405975-b2236ad6-8196-11e8-9988-8dab074fd174.png)

![image](https://i.imgur.com/ubUbUeV.gif)

![image](https://i.imgur.com/YIE57Dv.gif)

afterthought2 commented:
I don't really get it: you can still have two holsters with this, no? I'm not concerned about the balance, just the code aspects. What happens when I press h?

Alex6511 commented:
I like everything other than the holsters/webbings conflicting, outside of chest based holsters it doesn't make much sense.

@afterthought2 Same thing that happens if you have two now (by holding one in your hand) you get a menu that says which object you're pulling from.

Alex6511 commented:
@afterthought2 Same thing that happens if you have two now (by holding one in your hand) you get a menu that says which object you're pulling from.

afterthought2 commented:
Ah, OK. So long as it does something reasonable. Never had two holsters before myself.

CakeQ committed:
Belt Overhaul

more icons

fixes

fixes

extensifies holsters

tweaks

changes

fixes

fixes

fixes

makes utility belts brown

fixes

fixes

returns

CakeQ commented:
Map cleaned

CakeQ commented:
Funnily enough, this change brought to my attention the fact that lots of weapons have belt-specific onmob icons that are rarely ever seen in-game because everyone uses uniform-holsters. They looks great, though.

CakeQ commented:
PR close was a mis-click

babydoll commented:
Don't really personally like making all the belts boring and black

Also I guarantee that there's no way Psi will let this be merged with this amount of copy paste

Maveriknight commented:
I see this as proving problematic with exploration given their tendancy to end up with at least an E-gun when going to away sites with a high chance of danger. almost always swap the machete for a pistol holster in those scenarios.

comma commented on code/game/objects/items/weapons/storage/belt.dm:
```diff
@@ -30,14 +38,154 @@
+	holstered = null
+	SetName(initial(name))
+
+/obj/item/weapon/storage/belt/holster/proc/unholster(mob/user as mob)
```
Please tell me it's not a staright up copypaste of holster accessory code. Because it looks like a straight up copypaste of holster accessory code to me.

CakeQ replied:
what else am i supposed do smh

CrimsonShrike replied:
Probably some sort of extension would be in vogue. But may need someone to explain how-tos

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	src.sound_out = sound_out
+	src.can_holster = can_holster
+
+	atom_holder.verbs += /atom/proc/holster_verb
```
Need to remove this verb in `/Destroy()`

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	atom_holder = holder
+	src.storage = storage
+	src.sound_in = sound_in
+	src.sound_out = sound_out
```
Since you have default values this should be
```
src.sound_in = sound_in || src.sound_in
src.sound_out = sound_out || src.sound_out
```
Or they'll be nulled if not supplied.

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	var/sound_in = 'sound/effects/holster/holsterin.ogg'
+	var/sound_out = 'sound/effects/holster/holsterout.ogg'
+	var/list/can_holster = null
+	var/obj/item/holstered = null
```
You'll probably want to use a `/weakref` instead.
That way you don't have to track if/when `holstered` is deleted.
You still want to null this reference in `/Destroy()`

CakeQ replied:
Not doing this after what we discussed/implemented

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	if(!holstered)
+		return
+
+	if(istype(user.get_active_hand(),/obj) && istype(user.get_inactive_hand(),/obj))
```
Don't need to type check.
Just `if(user.get_active_hand() && user.get_inactive_hand())` will suffice.

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+			return 1
+	if(!holstered)
+		if(storage.storage_slots != null && storage.contents.len >= storage.storage_slots - 1)
+			to_chat(user, "<span class='notice'>\The [W] doesn't fit into \the [atom_holder]'s holster slot.</span>")
```
This message will come out twice, if `holster()` failed on the same check.

CakeQ replied:
holster() is never called if it gets to this part

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	set category = "Object"
+	set src in usr
+	if(!istype(usr, /mob/living)) return
+	if(usr.stat) return
```
`usr.incapacitated()`
This replaces both the `istype()` and `stat` check.

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	set src in usr
+	if(!istype(usr, /mob/living)) return
+	if(usr.stat) return
+	var/obj/item/clothing/accessory/storage/holster/holster = null
```
This can be moved to the usage on line 118.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	else
+		H.unholster(usr)
+
+/obj/item/weapon/storage/belt/holster/Initialize()
```
This `/Initialize()` as well as the one below should be moved to the respective type definitions.

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	storage.handle_item_insertion(holstered, 1)
+	holstered.add_fingerprint(user)
+	storage.w_class = max(storage.w_class, holstered.w_class)
+	user.visible_message("<span class='notice'>[user] holsters \the [holstered].</span>", "<span class='notice'>You holster \the [holstered].</span>")
```
Nice mix of using and not using \the

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+	holstered.add_fingerprint(user)
+	storage.w_class = max(storage.w_class, holstered.w_class)
+	user.visible_message("<span class='notice'>[user] holsters \the [holstered].</span>", "<span class='notice'>You holster \the [holstered].</span>")
+	atom_holder.name = "occupied [initial(atom_holder.name)]"
```
`atom_holder.SetName("occupied [initial(atom_holder.name)]")`

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+		clear_holster()
+		atom_holder.update_icon()
+
+/datum/extension/holster/proc/attempt_holster(obj/item/W as obj, mob/user as mob)
```
What exactly is this proc trying to do, that just calling `holster()` directly (and have it return TRUE/False on success/failure) wouldn't achieve?

CakeQ replied:
Needed to avoid having both "wont fit in holster" and "wont fit in storage" messages

PsiOmegaDelta replied:
Still a duplication of functionality.
Anyone who adds, removes, or alters these pre-checks will have to remember to do it in two places which is undesirable.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+			holster = locate() in S.accessories
+			H = get_extension(holster, /datum/extension/holster)
+	else
+		H = get_extension(src, /datum/extension/holster)
```
Because an atom can now potentially have multiple holsters with the removal of the holster slot, you probably want to add the following (or similar, this is written directly from my head, never compiled) 
``` 
/atom/proc/get_holsters()
    . = list()
    if(has_extension(src, /datum/extension/holster))
        .[name] = get_extension(src, /datum/extension/holster)

/obj/item/clothing/under/get_holsters()
    . = ..()
    var/holster_accessories_by_name = list()
    for(var/obj/accessory in accessories)
        if(has_extension(accessory, /datum/extension/holster))
            group_by(holster_accessories_by_name , accessory.name, accessory)

    for(var/accessory_name in holster_accessories_by_name)
        var/list/holster_accessories = holster_accessories_by_name[accessory_name]
        if(holster_accessories.len == 1)
            .[accessory_name] = get_extension(holster_accessory, /datum/extension/holster)
        else
            for(var/i = 1 to holster_accessories.len)
                var/holster_name = "[accessory_name] [i]"
                .[holster_name ] = get_extension(holster_accessory, /datum/extension/holster)
```
And change this proc to:
```
/atom/proc/holster_verb(var/holster_name in get_holsters())
[...]
    var/datum/extension/holster_extension/H = get_holsters()[holster_name]
    if(!H)
        return
[...]
```

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/modules/clothing/under/accessories/holster.dm:
```diff
@@ -1,148 +1,68 @@
+/obj/item/clothing/accessory/storage/holster/on_removed(mob/user as mob)
 	if(has_suit)
-		has_suit.verbs -= /obj/item/clothing/accessory/holster/verb/holster_verb
+		has_suit.verbs -= /atom/proc/holster_verb
```
Because you've removed the holster slot you'll need to check that there are no other accessories (besides itself) that also grant the `/atom/proc/holster_verb` verb.

BYOND doesn't add the same verb twice, so if two accessories do `has_suit.verbs += /atom/proc/holster_verb` then as soon as one of them do `has_suit.verbs -= /atom/proc/holster_verb` the verb is all gone.

PsiOmegaDelta replied:
Oh, and you probably want to check that the suit itself doesn't have the holster extension as well because that's a possibility now.

CakeQ replied:
Fixed.

PsiOmegaDelta commented on code/game/objects/items/weapons/storage/belt.dm:
```diff
@@ -17,11 +21,15 @@
 		M.update_inv_belt()
 
+	overlays.Cut()
+	if(overlay_flags |= BELT_OVERLAY_ITEMS)
```
You're adding (or-ing) BELT_OVERLAY_ITEMS to the `overlay_flags` var

PsiOmegaDelta replied:
What you want to do is `if(overlay_flags & BELT_OVERLAY_ITEMS)`

PsiOmegaDelta commented on code/game/objects/items/weapons/storage/belt.dm:
```diff
@@ -30,14 +38,64 @@
+
+	overlays.Cut()
+	var/datum/extension/holster/H = get_extension(src, /datum/extension/holster)
+	if(overlay_flags |= BELT_OVERLAY_ITEMS)
```
You're adding (or-ing) BELT_OVERLAY_ITEMS to the overlay_flags var

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,138 @@
+
+/obj/item/clothing/accessory/storage/holster/Initialize()
+	. = ..()
+	set_extension(hold, /datum/extension/holster, /datum/extension/holster, hold, sound_in, sound_out, can_holster)
```
Wrong extension target.
`set_extension(src, /datum/extension/holster, /datum/extension/holster, hold, sound_in, sound_out, can_holster)`

PsiOmegaDelta replied:
The same goes for all locations that do `get_extension(hold, ...)`

PsiOmegaDelta commented:
Are you claiming "Fixed" for your own convenience, because as far as I can tell you haven't done anything?

PsiOmegaDelta commented:
The finalized holster verb changes. As mentioned, this depends on `/obj/item/clothing/accessory/storage/holster` setting/getting the extension on itself, not `hold`.
```
/atom/proc/holster_verb(var/holster_name in get_holsters())
	set name = "Holster"
	set category = "Object"
	set src in usr
	
	if(usr.incapacitated())
		return
	
	var/datum/extension/holster/H = get_holsters()[holster_name]
	if(!H)
		return

	if(!H.holstered)
		var/obj/item/W = usr.get_active_hand()
		if(!istype(W, /obj/item))
			to_chat(usr, "<span class='warning'>You're not holding anything to holster.</span>")
			return
		H.holster(W, usr)
	else
		H.unholster(usr)
		
/atom/proc/get_holsters()
    . = list()
    if(has_extension(src, /datum/extension/holster))
        .[name] = get_extension(src, /datum/extension/holster)

/obj/item/clothing/under/get_holsters()
    . = ..()
    var/holster_accessories_by_name = list()
    for(var/obj/accessory in accessories)
        if(has_extension(accessory, /datum/extension/holster))
            group_by(holster_accessories_by_name , accessory.name, accessory)

    for(var/accessory_name in holster_accessories_by_name)
        var/list/holster_accessories = holster_accessories_by_name[accessory_name]
        if(holster_accessories.len == 1)
            .[accessory_name] = get_extension(holster_accessories[1], /datum/extension/holster)
        else
            for(var/i = 1 to holster_accessories.len)
                var/holster_name = "[accessory_name] [i]"
                .[holster_name] = get_extension(holster_accessories[i], /datum/extension/holster)
```

CakeQ commented:
Implemented your change requests

PsiOmegaDelta commented on code/game/objects/items/weapons/storage/belt.dm:
```diff
@@ -30,14 +38,68 @@
+	if(overlay_flags & BELT_OVERLAY_ITEMS)
+		for(var/obj/item/I in contents)
+			if(I == H.holstered)
+				if(overlay_flags & BELT_OVERLAY_HOLSTER)
```
Given the setup is almost seems like `BELT_OVERLAY_HOLSTER` might as well be defined as `(BELT_OVERLAY_ITEMS | 2)`, seeing how it serves no purpose without the first flag.

comma commented on code/game/objects/items/weapons/storage/belt.dm:
```diff
@@ -167,8 +301,43 @@
+		/obj/item/device/tape,
+		/obj/item/device/analyzer
 		)
+	can_holster = list(/obj/item/weapon/material/hatchet/machete, /obj/item/weapon/material/hatchet/machete/deluxe)
```
why subtype listed specificallly? Wouldn't just first thing work.

CakeQ replied:
not sure, I just copied over the listing from the machete holster accessory, which specifies both so I'd assume not

comma replied:
yet machete holster accessory only has main type. not filling me with confidence here.
one of them is wrong, check which.

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,144 @@
+	atom_holder.verbs += /atom/proc/holster_verb
+
+/datum/extension/holster/Destroy()
+	..()
```
Must return the result of ..() in `/Destroy()`

PsiOmegaDelta commented on code/datums/extensions/holster/holster.dm:
```diff
@@ -0,0 +1,144 @@
+
+/datum/extension/holster/proc/can_holster(var/obj/item/I)
+	if(can_holster)
+		if(is_type_in_list(I,can_holster))
```
Aren't these checks basically doing what `storage.handle_item_insertion()` does?

CakeQ replied:
no, holsters are unique in that they have their own dedicated slots within the storage

CakeQ commented:
@PsiOmegaDelta changed
